src/Microsoft.AspNetCore.Sockets/HttpConnectionDispatcher.cs

```diff
@@ -378,6 +359,28 @@ private ConnectionState CreateConnection(HttpContext context)
             }
         }
 
+        private async Task<IList<Message>> ExtractMessages(HttpContext context, byte[] buffer)
```

Comment:
You could also move the writing of the response to outside this method. Other methods do it this way, but I'm not sure I like it there either (even though I wrote it. That way you'd be doing the awaiting in the method that returns `Task` rather than `Task<T>` which **is** cacheable.